vcd/resource_vcd_nsxv_firewall_rule.go

```diff
@@ -0,0 +1,850 @@
+		return fmt.Errorf("unable to update firewall rule with ID %s: %s", d.Id(), err)
+	}
+
+	return resourceVcdNsxvFirewallRead(d, meta)
+}
+
+func resourceVcdNsxvFirewallRead(d *schema.ResourceData, meta interface{}) error {
+	vcdClient := meta.(*VCDClient)
+
+	edgeGateway, err := vcdClient.GetEdgeGatewayFromResource(d, "edge_gateway")
+	if err != nil {
+		return fmt.Errorf(errorUnableToFindEdgeGateway, err)
+	}
+
+	_, vdc, err := vcdClient.GetOrgAndVdcFromResource(d)
+	if err != nil {
+		return fmt.Errorf(errorRetrievingOrgAndVdc, err)
+	}
+
+	// Detect if this is data source or resource field and pick correct ID field
+	var isDatasource bool
+	id := d.Id()
+	if id == "" {
+		if ruleId, ok := d.GetOk("rule_id"); ok {
+			id = ruleId.(string)
+			isDatasource = true
+		}
+
+	}
+
+	readFirewallRule, err := edgeGateway.GetNsxvFirewallById(id)
+	if err != nil {
+		d.SetId("")
+		return fmt.Errorf("unable to find firewall rule with ID %s: %s", d.Id(), err)
+	}
+
+	err = setFirewallRuleData(d, readFirewallRule, edgeGateway, vdc)
+	if err != nil {
+		return err
+	}
+
+	if isDatasource {
+		d.SetId(readFirewallRule.ID)
+	}
+
+	return nil
+}
+
+func resourceVcdNsxvFirewallDelete(d *schema.ResourceData, meta interface{}) error {
+	vcdClient := meta.(*VCDClient)
+	vcdClient.lockParentEdgeGtw(d)
+	defer vcdClient.unLockParentEdgeGtw(d)
+
+	edgeGateway, err := vcdClient.GetEdgeGatewayFromResource(d, "edge_gateway")
+	if err != nil {
+		return fmt.Errorf(errorUnableToFindEdgeGateway, err)
+	}
+
+	err = edgeGateway.DeleteNsxvFirewallById(d.Id())
+	if err != nil {
+		return fmt.Errorf("error deleting firewall rule with id %s: %s", d.Id(), err)
+	}
+
+	d.SetId("")
+	return nil
+}
+
+// resourceVcdNsxvFirewallImport  is responsible for importing the resource.
+// The following steps happen as part of import
+// 1. The user supplies `terraform import _resource_name_ _the_id_string_` command
+// 2a. If the `_the_id_string_` contains a dot formatted path to resource as in the example below
+// it will try to import it. If it is found - the ID is set
+// 2b. If the `_the_id_string_` starts with `list@` and contains path to edge gateway similar to
+// `list@org.vdc.edge-gw` then the function lists all firewall rules and their IDs in that edge
+// gateway.
+// 2c. If the `_the_id_string_` does not match format described neither in '2a' nor in '2b' a
+// usage error message is printed
+//
+// Example resource name (_resource_name_): vcd_lb_nsxv_firewall_rule.my-test-fw-rule
+// Example import path (_the_id_string_): org.vdc.edge-gw.existing-firewall-rule-id
+func resourceVcdNsxvFirewallImport(d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
+	var commandOrgName, orgName, vdcName, edgeName, firewallRuleId string
+	var listRules, importRule bool
+
+	resourceURI := strings.Split(d.Id(), ".")
+	helpError := fmt.Errorf("resource name must be specified as " +
+		"'org.vdc.edge-gw.firewall-rule-id' or 'list@org.vdc.edge-gw' to get a list of rules")
```

Comment:
```suggestion
		"'org.vdc.edge-gw.firewall-rule-id' or 'list@org.vdc.edge-gw-name' to get a list of rules")
```

Reply:
@vbauzysvmware  either we use `-name` for each entity, or we leave it everywhere. Adding only one in the list is inconsistent.

Reply:
We agreed everthere add `-name` to be explicit

Reply:
+1 much more clearer for the user when scan reading and we have `-name` or `-id` explicitly.

Reply:
So do we make it `org-name.vdc-name.edge-gw-name.firewall-rule-id`?

Reply:
Agreed. Done.